persist/fs/seek_manager.go

```diff
@@ -137,75 +170,198 @@ func (m *seekerManager) CacheShardIndices(shards []uint32) error {
 	return multiErr.FinalError()
 }
 
-func (m *seekerManager) openAnyUnopenSeekers(byTime *seekersByTime) error {
-	start := m.earliestSeekableBlockStart()
-	end := m.latestSeekableBlockStart()
-	blockSize := m.namespaceMetadata.Options().RetentionOptions().BlockSize()
-	multiErr := xerrors.NewMultiError()
+func (m *seekerManager) ConcurrentIDBloomFilter(shard uint32, start time.Time) (*ManagedConcurrentBloomFilter, error) {
+	byTime := m.seekersByTime(shard)
 
-	for t := start; !t.After(end); t = t.Add(blockSize) {
-		tNano := xtime.ToUnixNano(t)
-		byTime.RLock()
-		_, exists := byTime.seekers[tNano]
-		byTime.RUnlock()
+	// Try fast RLock() first
+	byTime.RLock()
+	startNano := xtime.ToUnixNano(start)
+	seekersAndBloom, ok := byTime.seekers[startNano]
+	byTime.RUnlock()
 
-		if exists {
-			// Avoid opening a new seeker if already an open seeker
-			continue
-		}
+	if ok && seekersAndBloom.wg == nil {
+		return seekersAndBloom.bloomFilter, nil
+	}
 
-		byTime.Lock()
-		_, exists = byTime.seekers[tNano]
-		if exists {
-			byTime.Unlock()
-			continue
-		}
+	byTime.Lock()
+	seekersAndBloom, err := m.getOrOpenSeekersWithLock(startNano, byTime)
+	byTime.Unlock()
+	return seekersAndBloom.bloomFilter, err
+}
 
-		seeker, err := m.newOpenSeeker(byTime.shard, t)
-		if err != nil {
-			if err != errSeekerManagerFileSetNotFound {
-				// Best effort to open files, if not there don't bother opening
-				multiErr = multiErr.Add(err)
-			}
-			byTime.Unlock()
-			continue
+func (m *seekerManager) Borrow(shard uint32, start time.Time) (ConcurrentFileSetSeeker, error) {
+	byTime := m.seekersByTime(shard)
+
+	byTime.Lock()
+	// Track accessed to precache in open/close loop
+	byTime.accessed = true
+
+	startNano := xtime.ToUnixNano(start)
+	seekersAndBloom, err := m.getOrOpenSeekersWithLock(startNano, byTime)
+	if err != nil {
+		return nil, err
+	}
+
+	seekers := seekersAndBloom.seekers
+	availableSeekerIdx := -1
+	availableSeeker := borrowableSeeker{}
+	for i, seeker := range seekers {
+		if !seeker.isBorrowed {
+			availableSeekerIdx = i
+			availableSeeker = seeker
+			break
 		}
+	}
 
-		byTime.seekers[tNano] = seeker
+	// Should not occur in the case of a well-behaved caller
+	if availableSeekerIdx == -1 {
 		byTime.Unlock()
+		return nil, errNoAvailableSeekers
 	}
 
-	return multiErr.FinalError()
+	availableSeeker.isBorrowed = true
+	seekers[availableSeekerIdx] = availableSeeker
+	byTime.seekers[startNano] = seekersAndBloom
```

Comment:
Is this required to be set here?  It seems its not modified/changed since it was returned from ` m.getOrOpenSeekersWithLock(...)`.  We modify an entry in `seekers` but that should be copy into the existing slice.

Reply:
you're right its not, I'll remove it